docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,118 @@
+
+Created: Jan 23, 2023
+
+## Background
+
+Spin should have native support for using a variety of key-value (KV) datastores.
+
+Currently, users can use Redis for persistent key-value storage for Spin applications. This [documentation](https://developer.fermyon.com/cloud/data-redis.md) walks through how to use the existing Spin Redis SDK which leverages the [`outbound-redis.wit` interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/outbound-redis.wit). In contrast to the Redis interface, we are proposing a more general-purpose interface which can be used with a variety of implementations.
+
+## Proposal
+
+In order to support key-value stores, the following need to be added to Spin:
+
+- A `WIT` file that defines the key-value interface
+- SDK implementations for various programming languages
+- A default local key-value store
+
+Although not in scope for this proposal, we'll also want to expand the runtime configuration code added in [this PR](https://github.com/fermyon/spin/pull/798) to support configuring various key-value stores, including the default one.
+
+### Key-Value Interface (`.wit`)
+
+Spin should leverage the WebAssembly WASI subgroup’s work to define a universal key-value `WIT` interface. That work is taking place in [the `wasi-keyvalue` proposal](https://github.com/WebAssembly/wasi-keyvalue). However, the proposal was made assuming that `WIT` star imports are available, which is not yet the case. Therefore, we're modeling stores as pseudo-[resource handles](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource) which may be created using an `open` function and disposed using a `close` function.  Each operation on the store is a function which accepts a handle as its first parameter.
+
+Note that the syntax of the following `WIT` file matches the `wit-bindgen` version currently used by Spin, which is out-of-date with respect to the latest [`WIT` specification](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md) and implementation.  Once we're able to update `wit-bindgen`, we'll update the syntax of all the Spin `WIT` files, including this one.
+
+```fsharp
+// A handle to an open key-value store
+type store = u32
+
+// The set of errors which may be raised by functions in this interface
+variant error {
+  // Too many stores have been opened simultaneously. Closing one or more
+  // stores prior to retrying may address this.
+  store-table-full,
+
+  // The host does not recognize the store name requested.  Defining and
+  // configuring a store with that name in a runtime configuration file
+  // may address this.
+  no-such-store,
+
+  // The store handle provided is not recognized, i.e. it was either never
+  // opened or has been closed.
+  invalid-store,
+
+  // No key-value tuple exists for the specified key in the specified
+  // store.
+  no-such-key,
+
+  // Some implementation-specific error has occurred (e.g. I/O)
+  runtime(string)
```

Comment:
This label is a little weird since `no-such-key` (and arguably `store-table-full`) are also "runtime" errors. `spin-config.wit` uses `provider(string)`. I could also see `other(string)`.

Not a blocker. Naming is hard.

Reply:
`implementation` or `io`, perhaps?

Reply:
For the record, I've changed the name to `io`.